ffi/src/scan.rs

```diff
@@ -98,23 +116,39 @@ pub unsafe extern "C" fn scan(
     snapshot: Handle<SharedSnapshot>,
     engine: Handle<SharedExternEngine>,
     predicate: Option<&mut EnginePredicate>,
+    schema: Option<&mut EngineSchema>,
 ) -> ExternResult<Handle<SharedScan>> {
     let snapshot = unsafe { snapshot.clone_as_arc() };
-    scan_impl(snapshot, predicate).into_extern_result(&engine.as_ref())
+    scan_impl(snapshot, predicate, schema).into_extern_result(&engine.as_ref())
 }
 
 fn scan_impl(
     snapshot: SnapshotRef,
     predicate: Option<&mut EnginePredicate>,
+    schema: Option<&mut EngineSchema>,
 ) -> DeltaResult<Handle<SharedScan>> {
     let mut scan_builder = snapshot.scan_builder();
+
     if let Some(predicate) = predicate {
         let mut visitor_state = KernelExpressionVisitorState::default();
         let pred_id = (predicate.visitor)(predicate.predicate, &mut visitor_state);
         let predicate = unwrap_kernel_predicate(&mut visitor_state, pred_id);
         debug!("Got predicate: {:#?}", predicate);
         scan_builder = scan_builder.with_predicate(predicate.map(Arc::new));
     }
+
+    if let Some(schema) = schema {
+        let mut visitor_state = KernelSchemaVisitorState::default();
+        let schema_id = (schema.visitor)(schema.schema, &mut visitor_state);
+        let schema = unwrap_kernel_schema(&mut visitor_state, schema_id);
+        if schema.is_none() {
+            warn!("Passed a schema to scan, but visiting it did not produce a valid schema");
```

Comment:
hm - should we fail here instead?

Reply:
Advocating to push down the error to `unwrap_kernel_schema`: https://github.com/delta-io/delta-kernel-rs/pull/1463/files#r2558181987